# **Workshop "Data Ingestion with dlt": Homework**


---

## **Dataset & API**

We’ll use **NYC Taxi data** via the same custom API from the workshop:

🔹 **Base API URL:**  
```
https://us-central1-dlthub-analytics.cloudfunctions.net/data_engineering_zoomcamp_api
```
🔹 **Data format:** Paginated JSON (1,000 records per page).  
🔹 **API Pagination:** Stop when an empty page is returned.  

## **Question 1: dlt Version**

1. **Install dlt**:

In [1]:
 !pip install dlt[duckdb]

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 384.9/384.9 kB 8.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 144.8/144.8 kB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 884.3/884.3 kB 16.1 MB/s eta 0:00:00


> Or choose a different bracket—`bigquery`, `redshift`, etc.—if you prefer another primary destination. For this assignment, we’ll still do a quick test with DuckDB.

2. **Check** the version:


In [2]:
!dlt --version

dlt 1.6.1


or:

In [3]:
import dlt
print("dlt version:", dlt.__version__)

dlt version: 1.6.1


**Answer**:  
- Provide the **version** you see in the output. **dlt version: 1.6.1**

## **Question 2: Define & Run the Pipeline (NYC Taxi API)**

Use dlt to extract all pages of data from the API.

Steps:

1️⃣ Use the `@dlt.resource` decorator to define the API source.

2️⃣ Implement automatic pagination using dlt's built-in REST client.

3️⃣ Load the extracted data into DuckDB for querying.



In [4]:
import dlt
from dlt.sources.helpers.rest_client import RESTClient
from dlt.sources.helpers.rest_client.paginators import PageNumberPaginator

# Define the API resource for NYC taxi data
@dlt.resource(name="rides")
def ny_taxi():
    client = RESTClient(
        base_url="https://us-central1-dlthub-analytics.cloudfunctions.net",
        paginator=PageNumberPaginator(
            base_page=1,
            total_path=None
        )
    )

    for page in client.paginate("data_engineering_zoomcamp_api"):    # <--- API endpoint for retrieving taxi ride data
        yield page   # <--- yield data to manage memory

pipeline = dlt.pipeline(
    pipeline_name="ny_taxi_pipeline",
    destination="duckdb",
    dataset_name="ny_taxi_data"
)

Load the data into DuckDB to test:






In [5]:
load_info = pipeline.run(ny_taxi)
print(load_info)

Pipeline ny_taxi_pipeline load step completed in 3.29 seconds
1 load package(s) were loaded to destination duckdb and into dataset ny_taxi_data
The duckdb destination used duckdb:////content/ny_taxi_pipeline.duckdb location to store data
Load package 1739731819.1586761 is LOADED and contains no failed jobs


Start a connection to your database using native `duckdb` connection and look what tables were generated:

In [6]:
import duckdb
from google.colab import data_table
data_table.enable_dataframe_formatter()

# A database '<pipeline_name>.duckdb' was created in working directory so just connect to it

# Connect to the DuckDB database
conn = duckdb.connect(f"{pipeline.pipeline_name}.duckdb")

# Set search path to the dataset
conn.sql(f"SET search_path = '{pipeline.dataset_name}'")

# Describe the dataset
tables_metadata = conn.sql("DESCRIBE").df()

print("Number of tables in the DF:", len(tables_metadata))
print("Tables in the DF:")
tables_metadata

Number of tables in the DF: 4
Tables in the DF:


database        schema                 name  \
0  ny_taxi_pipeline  ny_taxi_data           _dlt_loads   
1  ny_taxi_pipeline  ny_taxi_data  _dlt_pipeline_state   
2  ny_taxi_pipeline  ny_taxi_data         _dlt_version   
3  ny_taxi_pipeline  ny_taxi_data                rides   

                                        column_names  \
0  [load_id, schema_name, status, inserted_at, sc...   
1  [version, engine_version, pipeline_name, state...   
2  [version, engine_version, inserted_at, schema_...   
3  [end_lat, end_lon, fare_amt, passenger_count, ...   

                                        column_types  temporary  
0  [VARCHAR, VARCHAR, BIGINT, TIMESTAMP WITH TIME...      False  
1  [BIGINT, BIGINT, VARCHAR, VARCHAR, TIMESTAMP W...      False  
2  [BIGINT, BIGINT, TIMESTAMP WITH TIME ZONE, VAR...      False  
3  [DOUBLE, DOUBLE, DOUBLE, BIGINT, VARCHAR, DOUB...      False

**Answer:**
* How many tables were created? **4 tables**

## **Question 3: Explore the loaded data**

Inspect the table `ride`:


In [7]:
df = pipeline.dataset(dataset_type="default").rides.df()

print("Total number of records:", len(df))
df.head(10)

Total number of records: 10000


end_lat    end_lon  fare_amt  passenger_count payment_type  start_lat  \
0  40.742963 -73.980072      45.0                1       Credit  40.641525   
1  40.740187 -74.005698       6.5                1       Credit  40.722065   
2  40.718043 -74.004745      12.5                5       Credit  40.761945   
3  40.739637 -73.985233       4.9                1         CASH  40.749802   
4  40.730032 -73.852693      25.7                1         CASH  40.776825   
5  40.777537 -73.976860       7.3                2       Credit  40.790582   
6  40.770277 -73.962125       3.7                1       Credit  40.767147   
7  40.774043 -73.951465       8.1                2         CASH  40.761750   
8  40.777985 -73.943683       6.1                1         CASH  40.766355   
9  40.720052 -74.009823       8.9                1         CASH  40.751327   

   start_lon  tip_amt  tolls_amt  total_amt  trip_distance  \
0 -73.787442      9.0       4.15      58.15          17.52   
1 -74.009767      1.0       0.00       8.50           1.56   
2 -73.983038      2.0       0.00      15.50           3.37   
3 -73.992247      0.0       0.00       5.40           1.11   
4 -73.949233      0.0       4.15      29.85          11.09   
5 -73.953652      2.0       0.00      10.30           2.10   
6 -73.966408      1.0       0.00       5.20           0.40   
7 -73.977773      0.0       0.00       8.60           2.24   
8 -73.959832      0.0       0.00       6.10           1.48   
9 -73.987588      0.0       0.00       9.90           2.72   

     trip_dropoff_date_time     trip_pickup_date_time  surcharge vendor_name  \
0 2009-06-14 23:48:00+00:00 2009-06-14 23:23:00+00:00        0.0         VTS   
1 2009-06-18 17:43:00+00:00 2009-06-18 17:35:00+00:00        1.0         VTS   
2 2009-06-10 18:27:00+00:00 2009-06-10 18:08:00+00:00        1.0         VTS   
3 2009-06-14 23:58:00+00:00 2009-06-14 23:54:00+00:00        0.5         VTS   
4 2009-06-13 13:23:00+00:00 2009-06-13 13:01:00+00:00        0.0         VTS   
5 2009-06-10 19:52:00+00:00 2009-06-10 19:43:00+00:00        1.0         VTS   
6 2009-06-10 20:09:00+00:00 2009-06-10 20:06:00+00:00        0.5         VTS   
7 2009-06-14 21:08:00+00:00 2009-06-14 20:57:00+00:00        0.5         VTS   
8 2009-06-14 12:56:00+00:00 2009-06-14 12:49:00+00:00        0.0         VTS   
9 2009-06-10 18:13:00+00:00 2009-06-10 18:03:00+00:00        1.0         VTS   

         _dlt_load_id         _dlt_id  store_and_forward  
0  1739731819.1586761  oLIziH7G6bXrzg                NaN  
1  1739731819.1586761  VxwueXf/P8zQkA                NaN  
2  1739731819.1586761  7cNIdi1+M4H8cA                NaN  
3  1739731819.1586761  nnAsLlK4akmlYQ                NaN  
4  1739731819.1586761  Kr+mZ9DiatVYmg                NaN  
5  1739731819.1586761  C4ZopZpTr9Jt+Q                NaN  
6  1739731819.1586761  xFjQfad/YEluoQ                NaN  
7  1739731819.1586761  lUGKjUZEZDre4g                NaN  
8  1739731819.1586761  GNk9XQWMkayMMw                NaN  
9  1739731819.1586761  /bJthdpRlNGzCw                NaN

**Answer:**
* What is the total number of records extracted? **10,000 records***

## **Question 4: Trip Duration Analysis**

Run the SQL query below to:

* Calculate the average trip duration in minutes.

In [10]:
with pipeline.sql_client() as client:
    res = client.execute_sql(
            """
            SELECT
            AVG(date_diff('minute', trip_pickup_date_time, trip_dropoff_date_time))
            FROM rides;
            """
        )
    # Prints column values of the first row
    avg_trip_duration = res[0][0]
    print("Average trip duration:", avg_trip_duration, "minutes")

Average trip duration: 12.3049 minutes


**Answer:**
* What is the average trip duration? **12.3049 minutes**